private/buf/bufgen/features.go

```diff
@@ -132,11 +134,22 @@ func checkRequiredFeatures(
 			sort.Slice(failedFeatures, func(i, j int) bool {
 				return failedFeatures[i] < failedFeatures[j]
 			})
+			// We only log failed features (whereas we error on failed editions). This is in
+			// keeping with CLI versions pre-1.32.0 (BSR-3931).
```

Comment:
Under the hood, proto3 optional fields are put into a synthetic `oneof`. As a result, the code generated by a plugin that does not support proto3 optional will still serialize correctly. 

Code generated by a plugin that does not support editions will most likely behave much worse. In practice, the plugin will likely print an error and exit with an error code (or populate `CodeGeneratorResponse.error`). But there is no guarantee that it does. It could just as well generate code that looks okay, but crashes at runtime, or serializes in a non-conformant way.

Even if we chose to keep `FEATURE_PROTO3_OPTIONAL` as a warning for BC, I think we need to error for `FEATURE_SUPPORTS_EDITIONS`.

Reply:
> I would really prefer to see a config setting to treat the error as a warning for a particular plugin+feature pair. That way we could make it default to being an error for v2 config (and buf migrate could auto-add the opt-out flag when migrating buf.gen.yaml, to preserve back compat).

I don't think that's too controversial, tbh, but it might take a little bit to implement, not necessarily for complexity reasons, but we want to make sure we capture a reasonable UX for the configs without breaking what we currently have released.

Given the comments in this thread, I'm going to take the route or warning on only on `FEATURE_PROTO3_OPTIONAL` for now, since I do think we want to move things closer to a safer model.